Code/Framework/AzCore/AzCore/Memory/AllocationRecords.cpp

```diff
@@ -369,20 +369,25 @@ namespace AZ::Debug
             return;
         }
 
-        Debug::AllocationInfo* ai{};
+        const auto [addressAlreadyRecorded, ai] = [this, address, newAddress]() -> std::pair<bool, Debug::AllocationInfo*>
         {
             AZStd::scoped_lock lock(m_recordsMutex);
             auto node = m_records.extract(address);
             if (node.empty())
             {
-                RegisterAllocation(newAddress, byteSize, alignment, stackSuppressCount);
-                return;
+                return {false, nullptr};
             }
 
             // Make a best effort to avoid reallocations from mutating the
             // records map when recording a reallocation
             node.key() = newAddress;
-            ai = &m_records.insert(AZStd::move(node)).position->second;
+            auto ai = &m_records.insert(AZStd::move(node)).position->second;
+            return {true, ai};
+        }();
+        if (!addressAlreadyRecorded)
+        {
+            RegisterAllocation(newAddress, byteSize, alignment, stackSuppressCount);
+            return;
```

Comment:
No, this was addressing a separate issue, where calling `realloc` with a `nullptr` as the original address would cause a deadlock, due to both `RegisterAllocation` and `RegisterReallocation` both attempting to acquire the `m_recordsMutex`.